packages/perseus/src/interactive2/movable.ts

```diff
@@ -239,41 +184,254 @@ _.extend(Movable.prototype, {
+            this.draw();
+        };
+
+        $(document).bind("vmousemove", moveHandler);
+        $(document).bind("vmouseup", upHandler);
+    }
+
+    _applyConstraints(
+        current: Coord,
+        previous: Coord,
+        extraOptions?: ConstraintCallbacks,
+    ) {
+        let skipRemaining = false;
+
+        return (this.state.constraints ?? []).reduce(
+            (memo: Coord | false, constraint) => {
+                // A move that has been cancelled won't be propagated to later
+                // constraints calls
+                if (memo === false) {
+                    return false;
+                }
+
+                if (skipRemaining) {
+                    return memo;
+                }
+
+                const result = constraint.call(this, memo, previous, {
+                    onSkipRemaining: () => {
+                        skipRemaining = true;
+                    },
+                    ...extraOptions,
+                });
+
+                if (result === false) {
+                    // Returning false cancels the move
+                    return false;
+                }
+                if (kpoint.is(result, 2)) {
+                    // Returning a coord from constraints overrides the move
+                    return result;
+                }
+                if (result === true || result == null) {
+                    // Returning true or undefined allow the move to occur
+                    return memo;
+                }
+                // Anything else is an error
+                throw new PerseusError(
+                    "Constraint returned invalid result: " + result,
+                    Errors.Internal,
+                );
+            },
+            current,
+        );
+    }
 
     // Change z-order to back
-    toBack: function () {
+    toBack() {
         if (this.state.mouseTarget) {
             this.state.mouseTarget.toBack();
         }
-    },
+    }
 
     // Change z-order to front
-    toFront: function () {
+    toFront() {
         if (this.state.mouseTarget) {
             this.state.mouseTarget.toFront();
         }
-    },
-});
+    }
 
-export default Movable;
+    /**
+     * Add a listener to any event: startMove, constraints, onMove, onMoveEnd,
+     * etc. If a listener is already bound to the given eventName and id, then
+     * it is overwritten by func.
+     *
+     * eventName: the string name of the event to listen to. one of:
+     *   "onMoveStart", "onMove", "onMoveEnd", "draw", "remove"
+     *
+     * id: a string id that can be used to remove this event at a later time
+     *   note: adding multiple listeners with the same id is undefined behavior
+     *
+     * func: the function to call when the event happens, which is called
+     *   with the event's standard parameters [usually (coord, prevCoord) or
+     *   (state, prevState)]
+     */
+    listen(eventName: string, id: string, func: () => unknown) {
```

Comment:
Potentially in a future PR, but it might be nice for the `eventName` to be a string union of eligible event names.